add bilinear weight generation

dmarsico1 opened the pull request:
Fixed the merge conflict from "monotone-holes" pull request.

An implementation of the (non-integrated) bilinear and generalized barycentric coordinate schemes for meshes with holes.  When the source mesh has no holes, the global dual of the source mesh is used to find a face whose nodes are source are source face centers that contains a given target point.  When the source mesh does have holes, we construct local dual faces around each source node to find such a face that contains a given target point.

vijaysm commented on src/LinearRemapFV.cpp:
```diff
@@ -1611,199 +1610,451 @@ void LinearRemapGeneralizedBarycentric(
+		
+	}
+	
+	else{
```
Can you add a comment that this is the else block for mesh with holes. It makes readability better

vijaysm commented on src/LinearRemapFV.cpp:
```diff
@@ -1611,199 +1610,451 @@ void LinearRemapGeneralizedBarycentric(
+			/* find points closest to the origin and within distance radius */
+			//struct *presults;
+										 
+			struct kdres *presults = kd_nearest_range3(kdTarget,faceFirstCenter.x, faceFirstCenter.y, 
```
You have to be somewhat careful for distance-based nearest neighbor queries to make sure it doesn't return points across a land boundary etc when there is a thin peninsula for example.

dmarsico1 replied:
Here, the distance-based nearest neighbor is used only to find a set of target faces whose centers could potentially be within a given source face.  Once we have that set of target faces, we then check each face to see whether its center is contained within that source face.  If the center is not contained within the given source face, it will be ignored.

vijaysm reviewed (changes requested):
I tested some of the bilinear variants on a mesh with holes. It produced maps without errors. I haven't done more rigorous testing yet. Once I finish testing with just TR mesh files, I will try to see if I can get a MOAB-TR test running as well this week.

I made several comments during the code review.

vijaysm commented:
I also evaluated the quality of the maps in comparison to a first-order conservative map. The workflow here is to generate maps for several combinations, given a full global source mesh (ATM) and a target mesh (OCN) with holes. Then I generated test data and applied the map to the same source data. I expected the integrals to match between FV and intbilin* variants. They are close but do not match to machine precision. Is this expected? @paullric @dmarsico1 

```
------------------------------------------------------------
Parameters:
  --map <string> ["map_weights_fv.nc"] 
  --in_data <string> ["sourcedata.nc"] 
  --out_data <string> ["targetdata_fv.nc"] 
  --in_data_list <string> [""] 
  --out_data_list <string> [""] 
  --var <string> [""] 
  --ncol_name <string> ["ncol"] 
  --bounds <string> [""] 
  --mesh_in <string> [""] 
  --mesh_out <string> [""] 
  --mesh_ov <string> [""] 
  --in_np <integer> [0] 
  --out_np <integer> [0] 
  --out_double <bool> [false] 
  --preserve <string> [""] 
  --preserveall <bool> [false] 
  --fillvalue <double> [0.000000] 
  --logdir <string> [""] 
  --concave <bool> [false] 
  --gll <bool> [false] 
------------------------------------------------------------
Loading offline map.. Done
Processing "sourcedata.nc"
..Psi
....Source Mass: 2.513274122871830e+01 Min 1.0050910271e+00 Max 2.9949089729e+00
....Target Mass: 1.641265842124411e+01 Min 0.0000000000e+00 Max 2.9949089729e+00
..Done
------------------------------------------------------------
Parameters:
  --map <string> ["map_weights_bilin.nc"] 
  --in_data <string> ["sourcedata.nc"] 
  --out_data <string> ["targetdata_bilin.nc"] 
  --in_data_list <string> [""] 
  --out_data_list <string> [""] 
  --var <string> [""] 
  --ncol_name <string> ["ncol"] 
  --bounds <string> [""] 
  --mesh_in <string> [""] 
  --mesh_out <string> [""] 
  --mesh_ov <string> [""] 
  --in_np <integer> [0] 
  --out_np <integer> [0] 
  --out_double <bool> [false] 
  --preserve <string> [""] 
  --preserveall <bool> [false] 
  --fillvalue <double> [0.000000] 
  --logdir <string> [""] 
  --concave <bool> [false] 
  --gll <bool> [false] 
------------------------------------------------------------
Loading offline map.. Done
Processing "sourcedata.nc"
..Psi
....Source Mass: 2.513274122871830e+01 Min 1.0050910271e+00 Max 2.9949089729e+00
....Target Mass: 1.780184562374376e+01 Min 1.0087217114e+00 Max 2.9903052083e+00
..Done
------------------------------------------------------------
Parameters:
  --map <string> ["map_weights_intbilin.nc"] 
  --in_data <string> ["sourcedata.nc"] 
  --out_data <string> ["targetdata_intbilin.nc"] 
  --in_data_list <string> [""] 
  --out_data_list <string> [""] 
  --var <string> [""] 
  --ncol_name <string> ["ncol"] 
  --bounds <string> [""] 
  --mesh_in <string> [""] 
  --mesh_out <string> [""] 
  --mesh_ov <string> [""] 
  --in_np <integer> [0] 
  --out_np <integer> [0] 
  --out_double <bool> [false] 
  --preserve <string> [""] 
  --preserveall <bool> [false] 
  --fillvalue <double> [0.000000] 
  --logdir <string> [""] 
  --concave <bool> [false] 
  --gll <bool> [false] 
------------------------------------------------------------
Loading offline map.. Done
Processing "sourcedata.nc"
..Psi
....Source Mass: 2.513274122871830e+01 Min 1.0050910271e+00 Max 2.9949089729e+00
....Target Mass: 1.640762375921785e+01 Min 0.0000000000e+00 Max 2.9897096239e+00
..Done
------------------------------------------------------------
Parameters:
  --map <string> ["map_weights_intbilingb.nc"] 
  --in_data <string> ["sourcedata.nc"] 
  --out_data <string> ["targetdata_intbilingb.nc"] 
  --in_data_list <string> [""] 
  --out_data_list <string> [""] 
  --var <string> [""] 
  --ncol_name <string> ["ncol"] 
  --bounds <string> [""] 
  --mesh_in <string> [""] 
  --mesh_out <string> [""] 
  --mesh_ov <string> [""] 
  --in_np <integer> [0] 
  --out_np <integer> [0] 
  --out_double <bool> [false] 
  --preserve <string> [""] 
  --preserveall <bool> [false] 
  --fillvalue <double> [0.000000] 
  --logdir <string> [""] 
  --concave <bool> [false] 
  --gll <bool> [false] 
------------------------------------------------------------
Loading offline map.. Done
Processing "sourcedata.nc"
..Psi
....Source Mass: 2.513274122871830e+01 Min 1.0050910271e+00 Max 2.9949089729e+00
....Target Mass: 1.640762928506532e+01 Min 0.0000000000e+00 Max 2.9897268034e+00
..Done
------------------------------------------------------------
Parameters:
  --map <string> ["map_weights_invdist.nc"] 
  --in_data <string> ["sourcedata.nc"] 
  --out_data <string> ["targetdata_invdist.nc"] 
  --in_data_list <string> [""] 
  --out_data_list <string> [""] 
  --var <string> [""] 
  --ncol_name <string> ["ncol"] 
  --bounds <string> [""] 
  --mesh_in <string> [""] 
  --mesh_out <string> [""] 
  --mesh_ov <string> [""] 
  --in_np <integer> [0] 
  --out_np <integer> [0] 
  --out_double <bool> [false] 
  --preserve <string> [""] 
  --preserveall <bool> [false] 
  --fillvalue <double> [0.000000] 
  --logdir <string> [""] 
  --concave <bool> [false] 
  --gll <bool> [false] 
------------------------------------------------------------
Loading offline map.. Done
Processing "sourcedata.nc"
..Psi
....Source Mass: 2.513274122871830e+01 Min 1.0050910271e+00 Max 2.9949089729e+00
....Target Mass: 1.640738396796232e+01 Min 0.0000000000e+00 Max 2.9877778096e+00
..Done
------------------------------------------------------------
Parameters:
  --map <string> ["map_weights_delaunay.nc"] 
  --in_data <string> ["sourcedata.nc"] 
  --out_data <string> ["targetdata_delaunay.nc"] 
  --in_data_list <string> [""] 
  --out_data_list <string> [""] 
  --var <string> [""] 
  --ncol_name <string> ["ncol"] 
  --bounds <string> [""] 
  --mesh_in <string> [""] 
  --mesh_out <string> [""] 
  --mesh_ov <string> [""] 
  --in_np <integer> [0] 
  --out_np <integer> [0] 
  --out_double <bool> [false] 
  --preserve <string> [""] 
  --preserveall <bool> [false] 
  --fillvalue <double> [0.000000] 
  --logdir <string> [""] 
  --concave <bool> [false] 
  --gll <bool> [false] 
------------------------------------------------------------
Loading offline map.. Done
Processing "sourcedata.nc"
..Psi
....Source Mass: 2.513274122871830e+01 Min 1.0050910271e+00 Max 2.9949089729e+00
....Target Mass: 1.780195045132307e+01 Min 1.0086884956e+00 Max 2.9903536240e+00
..Done
------------------------------------------------------------
```

paullric commented:
Why is the minimum on the target mesh zero in some cases?  Seems if the ocean mesh is a subset of the atmosphere mesh then all cell average values should be within the bounds from the source mesh.

> On Feb 16, 2023, at 9:05 AM, Vijay M ***@***.***> wrote:
> 
> 
> I also evaluated the quality of the maps in comparison to a first-order conservative map. The workflow here is to generate maps for several combinations, given a full global source mesh (ATM) and a target mesh (OCN) with holes. Then I generated test data and applied the map to the same source data. I expected the integrals to match between FV and intbilin* variants. They are close but do not match to machine precision. Is this expected? @paullric <https://github.com/paullric> @dmarsico1 <https://github.com/dmarsico1>
> ------------------------------------------------------------
> Parameters:
>   --map <string> ["map_weights_fv.nc"] 
>   --in_data <string> ["sourcedata.nc"] 
>   --out_data <string> ["targetdata_fv.nc"] 
>   --in_data_list <string> [""] 
>   --out_data_list <string> [""] 
>   --var <string> [""] 
>   --ncol_name <string> ["ncol"] 
>   --bounds <string> [""] 
>   --mesh_in <string> [""] 
>   --mesh_out <string> [""] 
>   --mesh_ov <string> [""] 
>   --in_np <integer> [0] 
>   --out_np <integer> [0] 
>   --out_double <bool> [false] 
>   --preserve <string> [""] 
>   --preserveall <bool> [false] 
>   --fillvalue <double> [0.000000] 
>   --logdir <string> [""] 
>   --concave <bool> [false] 
>   --gll <bool> [false] 
> ------------------------------------------------------------
> Loading offline map.. Done
> Processing "sourcedata.nc"
> ..Psi
> ....Source Mass: 2.513274122871830e+01 Min 1.0050910271e+00 Max 2.9949089729e+00
> ....Target Mass: 1.641265842124411e+01 Min 0.0000000000e+00 Max 2.9949089729e+00
> ..Done
> ------------------------------------------------------------
> Parameters:
>   --map <string> ["map_weights_bilin.nc"] 
>   --in_data <string> ["sourcedata.nc"] 
>   --out_data <string> ["targetdata_bilin.nc"] 
>   --in_data_list <string> [""] 
>   --out_data_list <string> [""] 
>   --var <string> [""] 
>   --ncol_name <string> ["ncol"] 
>   --bounds <string> [""] 
>   --mesh_in <string> [""] 
>   --mesh_out <string> [""] 
>   --mesh_ov <string> [""] 
>   --in_np <integer> [0] 
>   --out_np <integer> [0] 
>   --out_double <bool> [false] 
>   --preserve <string> [""] 
>   --preserveall <bool> [false] 
>   --fillvalue <double> [0.000000] 
>   --logdir <string> [""] 
>   --concave <bool> [false] 
>   --gll <bool> [false] 
> ------------------------------------------------------------
> Loading offline map.. Done
> Processing "sourcedata.nc"
> ..Psi
> ....Source Mass: 2.513274122871830e+01 Min 1.0050910271e+00 Max 2.9949089729e+00
> ....Target Mass: 1.780184562374376e+01 Min 1.0087217114e+00 Max 2.9903052083e+00
> ..Done
> ------------------------------------------------------------
> Parameters:
>   --map <string> ["map_weights_intbilin.nc"] 
>   --in_data <string> ["sourcedata.nc"] 
>   --out_data <string> ["targetdata_intbilin.nc"] 
>   --in_data_list <string> [""] 
>   --out_data_list <string> [""] 
>   --var <string> [""] 
>   --ncol_name <string> ["ncol"] 
>   --bounds <string> [""] 
>   --mesh_in <string> [""] 
>   --mesh_out <string> [""] 
>   --mesh_ov <string> [""] 
>   --in_np <integer> [0] 
>   --out_np <integer> [0] 
>   --out_double <bool> [false] 
>   --preserve <string> [""] 
>   --preserveall <bool> [false] 
>   --fillvalue <double> [0.000000] 
>   --logdir <string> [""] 
>   --concave <bool> [false] 
>   --gll <bool> [false] 
> ------------------------------------------------------------
> Loading offline map.. Done
> Processing "sourcedata.nc"
> ..Psi
> ....Source Mass: 2.513274122871830e+01 Min 1.0050910271e+00 Max 2.9949089729e+00
> ....Target Mass: 1.640762375921785e+01 Min 0.0000000000e+00 Max 2.9897096239e+00
> ..Done
> ------------------------------------------------------------
> Parameters:
>   --map <string> ["map_weights_intbilingb.nc"] 
>   --in_data <string> ["sourcedata.nc"] 
>   --out_data <string> ["targetdata_intbilingb.nc"] 
>   --in_data_list <string> [""] 
>   --out_data_list <string> [""] 
>   --var <string> [""] 
>   --ncol_name <string> ["ncol"] 
>   --bounds <string> [""] 
>   --mesh_in <string> [""] 
>   --mesh_out <string> [""] 
>   --mesh_ov <string> [""] 
>   --in_np <integer> [0] 
>   --out_np <integer> [0] 
>   --out_double <bool> [false] 
>   --preserve <string> [""] 
>   --preserveall <bool> [false] 
>   --fillvalue <double> [0.000000] 
>   --logdir <string> [""] 
>   --concave <bool> [false] 
>   --gll <bool> [false] 
> ------------------------------------------------------------
> Loading offline map.. Done
> Processing "sourcedata.nc"
> ..Psi
> ....Source Mass: 2.513274122871830e+01 Min 1.0050910271e+00 Max 2.9949089729e+00
> ....Target Mass: 1.640762928506532e+01 Min 0.0000000000e+00 Max 2.9897268034e+00
> ..Done
> ------------------------------------------------------------
> Parameters:
>   --map <string> ["map_weights_invdist.nc"] 
>   --in_data <string> ["sourcedata.nc"] 
>   --out_data <string> ["targetdata_invdist.nc"] 
>   --in_data_list <string> [""] 
>   --out_data_list <string> [""] 
>   --var <string> [""] 
>   --ncol_name <string> ["ncol"] 
>   --bounds <string> [""] 
>   --mesh_in <string> [""] 
>   --mesh_out <string> [""] 
>   --mesh_ov <string> [""] 
>   --in_np <integer> [0] 
>   --out_np <integer> [0] 
>   --out_double <bool> [false] 
>   --preserve <string> [""] 
>   --preserveall <bool> [false] 
>   --fillvalue <double> [0.000000] 
>   --logdir <string> [""] 
>   --concave <bool> [false] 
>   --gll <bool> [false] 
> ------------------------------------------------------------
> Loading offline map.. Done
> Processing "sourcedata.nc"
> ..Psi
> ....Source Mass: 2.513274122871830e+01 Min 1.0050910271e+00 Max 2.9949089729e+00
> ....Target Mass: 1.640738396796232e+01 Min 0.0000000000e+00 Max 2.9877778096e+00
> ..Done
> ------------------------------------------------------------
> Parameters:
>   --map <string> ["map_weights_delaunay.nc"] 
>   --in_data <string> ["sourcedata.nc"] 
>   --out_data <string> ["targetdata_delaunay.nc"] 
>   --in_data_list <string> [""] 
>   --out_data_list <string> [""] 
>   --var <string> [""] 
>   --ncol_name <string> ["ncol"] 
>   --bounds <string> [""] 
>   --mesh_in <string> [""] 
>   --mesh_out <string> [""] 
>   --mesh_ov <string> [""] 
>   --in_np <integer> [0] 
>   --out_np <integer> [0] 
>   --out_double <bool> [false] 
>   --preserve <string> [""] 
>   --preserveall <bool> [false] 
>   --fillvalue <double> [0.000000] 
>   --logdir <string> [""] 
>   --concave <bool> [false] 
>   --gll <bool> [false] 
> ------------------------------------------------------------
> Loading offline map.. Done
> Processing "sourcedata.nc"
> ..Psi
> ....Source Mass: 2.513274122871830e+01 Min 1.0050910271e+00 Max 2.9949089729e+00
> ....Target Mass: 1.780195045132307e+01 Min 1.0086884956e+00 Max 2.9903536240e+00
> ..Done
> ------------------------------------------------------------
> —
> Reply to this email directly, view it on GitHub <https://github.com/ClimateGlobalChange/tempestremap/pull/112#issuecomment-1433417457>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABIVPZHM4V3ZT462JTL4BFTWXZM5JANCNFSM6AAAAAATSRBWWE>.
> You are receiving this because you were mentioned.
>

vijaysm commented:
That is my other question as well, even for FV maps.

paullric commented:
Is it possible that the overlap mesh is missing a target face?  The bilinear map doesn’t use the overlap mesh, so it makes sense that it wouldn’t suffer from that issue.

~ Paul

> On Feb 17, 2023, at 8:04 AM, Vijay M ***@***.***> wrote:
> 
> 
> That is my other question as well, even for FV maps.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/ClimateGlobalChange/tempestremap/pull/112#issuecomment-1434851574>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABIVPZHDFKYP5GDZAS7XC2DWX6OP5ANCNFSM6AAAAAATSRBWWE>.
> You are receiving this because you were mentioned.
>

dmarsico1 commented:
The bilinear maps won't give the same results as the first order conservative FV maps, as the bilinear map is not conservative.  As for the 0.0 minimum values on the target mesh, that is strange, and it probably does have something to do with the overlap mesh. @vijaysm  @paullric

vijaysm commented:
I added the shell script and the input/output meshes I used for the above comparison tests. Feel free to replicate on your end. After you change the TempestRemap executable locations, you should be able to just run the script.

[1] https://www.dropbox.com/sh/apebzb18x79wpcc/AADWIZxYIx52LsSiUD59A65na?dl=0

vijaysm commented:
Any updates on the above? @dmarsico1 @paullric Let me know when the review requests are addressed and I can retest things again. This week, I will test the branch with MOAB in parallel as well to see if it resolves the issues I saw previously.

dmarsico1 commented:
No new updates yet.  I've reviewed the changes that were requested on the pull request, and they all seem doable.  I haven't figured out why the minimum is 0.0 yet, but I'll let you know when I do. @paullric, @vijaysm

vijaysm commented:
Ok sounds good. As mentioned previously, the files you need to reproduce the issue are in that Dropbox link.

dmarsico1 commented:
Just a quick update regarding the zero field values @paullric @vijaysm.  I think the issue is that the overlap mesh is missing some target faces.  I swapped the input and output meshes in the overlap mesh generation, and that seemed to fix the issue in that the minimum field value is no longer zero. 

Try this command for the overlap mesh:

GenerateOverlapMesh --b ATMSourceMesh.g --a  recMeshOcn.TargetMesh.g --out overlap.g

paullric commented:
Admittedly the default algorithm is not very robust.

~ Paul

> On Mar 18, 2023, at 5:36 PM, David Marsico ***@***.***> wrote:
> 
> 
> Just a quick update regarding the zero field values @paullric <https://github.com/paullric> @vijaysm <https://github.com/vijaysm>. I think the issue is that the overlap mesh is missing some target faces. I swapped the input and output meshes in the overlap mesh generation, and that seemed to fix the issue in that the minimum field value is no longer zero.
> 
> Try this command for the overlap mesh:
> 
> GenerateOverlapMesh --b ATMSourceMesh.g --a recMeshOcn.TargetMesh.g --out overlap.g
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/ClimateGlobalChange/tempestremap/pull/112#issuecomment-1475049058>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABIVPZBETFWVSGRHDLPH74DW4ZIIDANCNFSM6AAAAAATSRBWWE>.
> You are receiving this because you were mentioned.
>

rljacob commented:
@paullric what's the status on getting this merged?

rljacob commented:
We are using this in our E3SM moab coupler testing and it works if you increase the tolerance.

rljacob commented:
@dmarsico1 could you change the title to something more descriptive like "add bilinear weight generation".  Also add more info to the PR description.

paullric commented:
> @paullric what's the status on getting this merged?

@dmarsico1 and I will discuss this Friday then merge, assuming no issues

rljacob commented:
There are a bunch of comments from Vijay that should be addressed.

rljacob commented:
See also https://github.com/dmarsico1/tempestremap/pull/2

dmarsico1 commented on src/FiniteVolumeTools.cpp:
```diff
@@ -1698,3 +1696,293 @@ void InvertFitArray_LeastSquares(
+	std::vector<double> & vecWeights,
+	std::vector<int> & vecContributingFaces
+) {
+	
```
vecWeights and vecContributingFaces are now resized inside the method as needed.

dmarsico1 commented on src/FiniteVolumeTools.cpp:
```diff
@@ -1698,3 +1696,293 @@ void InvertFitArray_LeastSquares(
+	NodeVector & nodesFaceI,
+	std::vector<double> & vecWeights
+) {
+			
```
The size of vecWeights is now asserted to be equal to the size of nodesFaceI.

dmarsico1 commented on src/FiniteVolumeTools.cpp:
```diff
@@ -877,13 +876,12 @@ void TriangleLineIntersection(
 	NodeVector & nodesP,
-	DataArray1D<double> & dCoeffs,
-	bool & fConverged
+	DataArray1D<double> & dCoeffs
```
In practice, the scheme never failed to converge, so I removed the flag.  I still kept in the exception if the matrix is not invertible.

dmarsico1 commented on src/FiniteVolumeTools.cpp:
```diff
@@ -1698,3 +1696,293 @@ void InvertFitArray_LeastSquares(
+			
+			double dMagNodeDiff = nodeDiff.Magnitude();	
+			
+			if ( dMagNodeDiff < 1e-12 ){
```
I changed this to 1e-08 for the bilinear and generalized barycentric weights (see our email exchange about making this the default).

dmarsico1 commented on src/FiniteVolumeTools.cpp:
```diff
@@ -1698,3 +1696,293 @@ void InvertFitArray_LeastSquares(
+		
+		for (int i = 0; i < 3; i++){
+			
+			Node nodeI = nodesFaceI[i];
```
This has been fixed.

dmarsico1 commented:
@vijaysm @paullric I finished addressing all of the comments, and the code for the (non-integrated) bilinear weight generation should be ready to merge.

vijaysm commented:
Thanks @dmarsico1 I will take another look at the changes and approve if it looks good. Last I checked intbilin* was't working either for the meshes I provided above, but we can address that in a separate PR next.

dmarsico1 commented:
@vijaysm Sounds good. I'm currently working on the intbilin and intbilingb variants, and will do a separate pull request for those.  I was just going to make sure there were no more issues with the bilin variants.

vijaysm commented:
Thanks for pushing all the changes. Bilinear looks good now. @dmarsico1

Please go ahead and approve/merge the PR. @paullric

rljacob commented:
@paullric can this get merged?

rljacob commented:
@dmarsico1 why did you merge master into your branch?

dmarsico1 commented:
@rljacob I needed the function "EnforceConsistency()" to implement the integrated methods intbilin and intbilingb (for these methods, the remapping operator will not be consistent for faces that are only partially covered by the overlap mesh, i.e. near coastlines, so there needs to be a way of enforcing consistency for these faces).

rljacob commented:
Ok. So you're going to continue to develop those on this monotone-holes2 branch?    It would be a cleaner development history if you started a new branch from master after this one was merged.

vijaysm commented:
@paullric this PR is good to go. I have further comments for intbilin* but that can be raised as a separate issue. We need the bilinear near term for the online MOAB-E3SM runs.
